File: src/grid.cpp
#include <vector>
#include <string>
#include <iostream>
#include <cmath>
#include "grid.h"
using namespace std; 

vector<vector<int>> Grid::convertTo2DVector(const vector<int>& grid) {
    int size = sqrt(grid.size());
    vector<vector<int>> temp(size, vector<int>(size));
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            temp[i][j] = grid[i * size + j];
        }
    }
    return temp;
}

vector<vector<int>> Grid::convertTo2DVector(const string& grid) {
    int size = sqrt(grid.size());
    if (size * size != grid.size()) {
        throw invalid_argument("String length must be a perfect square. String length: " + to_string(grid.size()));
    }
    vector<vector<int>> temp(size, vector<int>(size));
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            temp[i][j] = grid[i * size + j] - '0'; // 假设字符串中的每个字符是一个数字
        }
    }
    return temp;
}

Grid::Grid(int grid_size, int box_size){
    if (int(sqrt(box_size)) * int(sqrt(box_size)) != box_size) {
        throw invalid_argument("Box size must be a perfect square. Box size: " + to_string(box_size));
    }
    if (grid_size % int(sqrt(box_size)) != 0) {
        throw invalid_argument("Grid size must be divisible by box len. Grid size: " + to_string(grid_size) + " Box size: " + to_string(box_size) + " Box len: " + to_string(int(sqrt(box_size))));
    }
    this->GRID_SIZE = grid_size;
    this->BOX_SIZE = box_size;
    this->BOX_LEN = sqrt(box_size);
    this->grid = vector<vector<int>>(grid_size, vector<int>(grid_size));
}

Grid::Grid(const string& grid): Grid(convertTo2DVector(grid)){}

Grid::Grid(const vector<int>& grid): Grid(convertTo2DVector(grid)){}

Grid::Grid(const vector<vector<int>>& grid){
    if (grid.size() != grid[0].size()) {
        throw invalid_argument("Grid must be square. Rows: " + to_string(grid.size()) + " Columns: " + to_string(grid[0].size()));
    }
    this->GRID_SIZE = grid.size();
    this->BOX_SIZE = this->GRID_SIZE;
    this->BOX_LEN = sqrt(this->GRID_SIZE);
    this->grid = grid;
}

Grid::Grid(const Grid& grid){
    this->GRID_SIZE = grid.GRID_SIZE;
    this->BOX_SIZE = grid.BOX_SIZE;
    this->BOX_LEN = grid.BOX_LEN;
    this->grid = grid.grid;
}

Grid::~Grid(){

}

bool Grid::operator==(const Grid& grid){
    if (this->GRID_SIZE != grid.GRID_SIZE) {
        return false;
    }
    if (this->BOX_SIZE != grid.BOX_SIZE) {
        return false;
    }
    if (this->BOX_LEN != grid.BOX_LEN) {
        return false;
    }
    for (int i = 0; i < GRID_SIZE; i++)
    {
        for (int j = 0; j < GRID_SIZE; j++)
        {
            if (this->grid[i][j] != grid.grid[i][j]) {
                return false;
            }
        }
    }
    return true;
}

vector<int> Grid::getRow(int row){
    return grid[row];
}

vector<int> Grid::getColumn(int col){
    for (int i = 0; i < GRID_SIZE; i++)
    {
        cout << grid[i][col] << endl;
    }
}

int Grid::getBox(int row, int col){
    return grid[row][col];
}

bool Grid::setRow(int row, const vector<int>& values){
    if (row >= GRID_SIZE || row < 0) {
        throw invalid_argument("Row index out of range. Row: " + to_string(row));
        return false;
    }
    if (values.size() != GRID_SIZE) {
        throw invalid_argument("Values size must be equal to grid size. Values size: " + to_string(values.size()));
        return false;
    }
    grid[row] = values;
    return true;
}

bool Grid::setColumn(int col, const vector<int>& values){
    if (col >= GRID_SIZE || col < 0) {
        throw invalid_argument("Column index out of range. Column: " + to_string(col));
        return false;
    }
    if (values.size() != GRID_SIZE) {
        throw invalid_argument("Values size must be equal to grid size. Values size: " + to_string(values.size()));
        return false;
    }
    for (int i = 0; i < GRID_SIZE; i++)
    {
        grid[i][col] = values[i];
    }
    return true;
}

bool Grid::setBox(int row, int col, int value){
    if (row >= GRID_SIZE || row < 0) {
        throw invalid_argument("Row index out of range. Row: " + to_string(row));
        return false;
    }
    if (col >= GRID_SIZE || col < 0) {
        throw invalid_argument("Column index out of range. Column: " + to_string(col));
        return false;
    }
    grid[row][col] = value;
    return true;
}

bool Grid::setGrid(const vector<vector<int>>& grid){
    if (grid.size() != GRID_SIZE) {
        throw invalid_argument("Grid size must be equal to grid size. Grid size: " + to_string(grid.size()));
        return false;
    }
    for (int i = 0; i < GRID_SIZE; i++)
    {
        if (grid[i].size() != GRID_SIZE) {
            throw invalid_argument("Grid size must be equal to grid size. Grid size: " + to_string(grid[i].size()));
            return false;
        }
    }
    this->grid = grid;
    return true;
}

void Grid::printGrid(){
    for (int i = 0; i < GRID_SIZE; i++)
    {
        for (int j = 0; j < GRID_SIZE; j++)
        {
            cout << grid[i][j] << " ";
        }
        cout << endl;
    }
}

File: python/grid.py
import math

class Grid:
    def __init__(self, grid_size=None, box_size=None, grid=None):
        if grid_size is not None and box_size is not None:
            if int(math.sqrt(box_size)) * int(math.sqrt(box_size)) != box_size:
                raise ValueError(f"Box size must be a perfect square. Box size: {box_size}")
            if grid_size % int(math.sqrt(box_size)) != 0:
                raise ValueError(f"Grid size must be divisible by box len. Grid size: {grid_size} Box size: {box_size} Box len: {int(math.sqrt(box_size))}")
            self.GRID_SIZE = grid_size
            self.BOX_SIZE = box_size
            self.BOX_LEN = int(math.sqrt(box_size))
            self.grid = [[0] * grid_size for _ in range(grid_size)]
        elif isinstance(grid, str):
            self.grid = self.convertTo2DVector(grid)
            self.GRID_SIZE = len(self.grid)
            self.BOX_SIZE = self.GRID_SIZE
            self.BOX_LEN = int(math.sqrt(self.GRID_SIZE))
        elif isinstance(grid, list):
            if all(isinstance(i, int) for i in grid):
                self.grid = self.convertTo2DVector(grid)
            elif all(isinstance(i, list) for i in grid):
                if len(grid) != len(grid[0]):
                    raise ValueError(f"Grid must be square. Rows: {len(grid)} Columns: {len(grid[0])}")
                self.grid = grid
            else:
                raise TypeError("Grid must be a string or a list of integers or a 2D list of integers.")
            self.GRID_SIZE = len(self.grid)
            self.BOX_SIZE = self.GRID_SIZE
            self.BOX_LEN = int(math.sqrt(self.GRID_SIZE))
        else:
            raise TypeError("Invalid arguments for Grid initialization.")
        
    def __copy__(self):
        new_grid = Grid(grid_size=self.GRID_SIZE, box_size=self.BOX_SIZE)
        new_grid.grid = self.grid.copy()
        return new_grid

    def __deepcopy__(self, memo):
        new_grid = Grid(grid_size=self.GRID_SIZE, box_size=self.BOX_SIZE)
        new_grid.grid = copy.deepcopy(self.grid, memo)
        return new_grid
    
    @staticmethod
    def convertTo2DVector(grid):
        if isinstance(grid, list):
            size = int(math.sqrt(len(grid)))
            return [[grid[i * size + j] for j in range(size)] for i in range(size)]
        elif isinstance(grid, str):
            size = int(math.sqrt(len(grid)))
            if size * size != len(grid):
                raise ValueError(f"String length must be a perfect square. String length: {len(grid)}")
            return [[int(grid[i * size + j]) for j in range(size)] for i in range(size)]
        else:
            raise TypeError("Grid must be a string or a list of integers.")

    def __eq__(self, other):
        if not isinstance(other, Grid):
            return False
        return self.GRID_SIZE == other.GRID_SIZE and self.BOX_SIZE == other.BOX_SIZE and self.BOX_LEN == other.BOX_LEN and self.grid == other.grid

    def getRow(self, row):
        return self.grid[row]

    def getColumn(self, col):
        return [self.grid[i][col] for i in range(self.GRID_SIZE)]

    def getBox(self, row, col):
        return self.grid[row][col]

    def setRow(self, row, values):
        if row >= self.GRID_SIZE or row < 0:
            raise ValueError(f"Row index out of range. Row: {row}")
        if len(values) != self.GRID_SIZE:
            raise ValueError(f"Values size must be equal to grid size. Values size: {len(values)}")
        self.grid[row] = values
        return True

    def setColumn(self, col, values):
        if col >= self.GRID_SIZE or col < 0:
            raise ValueError(f"Column index out of range. Column: {col}")
        if len(values) != self.GRID_SIZE:
            raise ValueError(f"Values size must be equal to grid size. Values size: {len(values)}")
        for i in range(self.GRID_SIZE):
            self.grid[i][col] = values[i]
        return True

    def setBox(self, row, col, value):
        if row >= self.GRID_SIZE or row < 0:
            raise ValueError(f"Row index out of range. Row: {row}")
        if col >= self.GRID_SIZE or col < 0:
            raise ValueError(f"Column index out of range. Column: {col}")
        self.grid[row][col] = value
        return True

    def setGrid(self, grid):
        if len(grid) != self.GRID_SIZE:
            raise ValueError(f"Grid size must be equal to grid size. Grid size: {len(grid)}")
        for row in grid:
            if len(row) != self.GRID_SIZE:
                raise ValueError(f"Grid size must be equal to grid size. Grid size: {len(row)}")
        self.grid = grid
        return True

    def printGrid(self):
        for row in self.grid:
            print(" ".join(map(str, row)))
